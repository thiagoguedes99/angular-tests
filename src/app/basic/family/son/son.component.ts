import { Component, EventEmitter, Input, Output } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-son',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './son.component.html',
  styleUrl: './son.component.scss'
})
export class SonComponent {
  @Input() client?: Client;
  @Output() onDeleteClient  = new EventEmitter();
  @Output() onClientUpdated = new EventEmitter<Client>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange( id: number ) {
    if ( !this.client ) return;

    // this.client.id = newId; // no usar
    this.client = { ...this.client, id };

    this.onClientUpdated.emit({ ...this.client })

  }
}
