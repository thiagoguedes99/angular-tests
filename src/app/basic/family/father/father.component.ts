import { Component } from '@angular/core';
import { JsonPipe } from '@angular/common';

import { SonComponent } from '../son/son.component';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-father',
  standalone: true,
  imports: [SonComponent, JsonPipe],
  templateUrl: './father.component.html',
  styleUrl: './father.component.scss'
})
export class FatherComponent {
  public client?: Client;

  onSetClient( name: string ) {
    this.client = { id: 1, name };
  }

}
