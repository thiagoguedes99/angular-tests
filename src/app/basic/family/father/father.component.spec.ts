import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherComponent } from './father.component';
import { By } from '@angular/platform-browser';
import { SonComponent } from '../son/son.component';

describe('FatherComponent', () => {
  let component: FatherComponent;
  let fixture: ComponentFixture<FatherComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [FatherComponent]
    })
    .compileComponents();

    fixture = TestBed.createComponent(FatherComponent);
    component = fixture.componentInstance;
    compiled = fixture.nativeElement;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should to match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should to render the client', () => {
    component.client = { id: 3, name: 'mock_thiago' };
    fixture.detectChanges();

    const codeDiv = compiled.querySelector('.mt-2');

    expect(codeDiv?.textContent).toContain('3');
    expect(codeDiv?.textContent).toContain('mock_thiago');
  });

  it('should to call onSetClient and update client', () => {
    component.client = { id: 3, name: 'mock_thiago' };
    expect(component.client).toEqual({ id: 3, name: 'mock_thiago' });

    component.onSetClient('other_name');
    expect(component.client).toEqual({ id: 1, name: 'other_name' });
  });

  it('should to delete the client when to emit onDeleteClient', () => {
    component.client = { id: 3, name: 'mock_thiago' };
    fixture.detectChanges();
    expect(component.client).toEqual({ id: 3, name: 'mock_thiago' });

    const sonDebugElement = fixture.debugElement.query( By.directive(SonComponent) );
    const sonComponent: SonComponent = sonDebugElement.componentInstance;

    sonComponent.onDeleteClient.emit();

    expect(component.client).toBe(undefined);
  });

  it('should to update the client when to emit onClientUpdated', () => {
    component.client = { id: 3, name: 'mock_thiago' };
    fixture.detectChanges();
    expect(component.client).toEqual({ id: 3, name: 'mock_thiago' });

    const sonDebugElement = fixture.debugElement.query( By.directive(SonComponent) );
    const sonComponent: SonComponent = sonDebugElement.componentInstance;

    sonComponent.onClientUpdated.emit({ id: 4, name: 'mock_thiago_2' });

    expect(component.client).toEqual({ id: 4, name: 'mock_thiago_2' });
  });
});
