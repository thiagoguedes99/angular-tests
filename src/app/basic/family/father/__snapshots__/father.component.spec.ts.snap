// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherComponent should to match snapshot 1`] = `
<div
  id="root1"
>
  <h1>
    Padre
  </h1>
  <button
    class="btn btn-primary"
  >
     Cargar Cliente

  </button>
  <br />
  <br />
  <code
    class="mt-2"
  >
    
  </code>
  <hr />
  <app-son>
    <h5>
      Hijo
    </h5>
    <code>
      
    </code>
    <br />
    <br />
    
    
  </app-son>
</div>
`;
